src/protocol/sender.rs

```diff
@@ -0,0 +1,622 @@
+
+                self.begin_transmit(time, slice, location);
+
+                location = PacketLocation::empty();
+                payload = payload.slice(0..self.max_packet_size)
+            } else {
+                location = location | PacketLocation::LAST;
+                self.begin_transmit(time, payload, location);
+                break;
+            }
+        }
+    }
+
+    fn begin_transmit(&mut self, time: Instant, payload: Bytes, location: PacketLocation) {
+        let packet = DataPacket {
+            dest_sockid: self.remote_sockid,
+            in_order_delivery: false, // TODO: research this
+            message_loc: location,
+            // if this marks the beginning of the next message, get a new message number, else don't
+            message_number: if location == PacketLocation::FIRST {
+                self.get_new_message_number()
+            } else {
+                self.latest_message_number()
+            },
+            seq_number: self.get_new_sequence_number(),
+            timestamp: self.time_base.timestamp_from(time),
+            payload,
+        };
+
+        self.buffer.push_back(packet)
+    }
+
+    pub fn pop_front(&mut self) -> Option<DataPacket> {
+        self.buffer.pop_front()
+    }
+
+    pub fn front(&self) -> Option<&DataPacket> {
+        self.buffer.front()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    pub fn latest_message_number(&self) -> MsgNumber {
+        self.next_message_number - 1
+    }
+
+    pub fn latest_seqence_number(&self) -> SeqNumber {
+        self.next_seq_number - 1
+    }
+
+    pub fn next_seqence_number(&self) -> SeqNumber {
+        self.next_seq_number
+    }
+
+    pub fn timestamp_now(&self) -> TimeStamp {
+        self.time_base.timestamp_now()
+    }
+
+    /// Gets the next available message number
+    fn get_new_message_number(&mut self) -> MsgNumber {
+        self.next_message_number += 1;
+        self.next_message_number - 1
+    }
+
+    /// Gets the next avilabe packet sequence number
+    fn get_new_sequence_number(&mut self) -> SeqNumber {
+        // this does looping for us
+        self.next_seq_number += 1;
+        self.next_seq_number - 1
+    }
+}
+
+struct SendBuffer {
+    /// The buffer to store packets for retransmision, sorted chronologically
+    buffer: VecDeque<DataPacket>,
+
+    /// The first sequence number in buffer, so seq number i would be found at
+    /// buffer[i - first_seq]
+    first_seq: SeqNumber,
+}
+
+impl SendBuffer {
+    pub fn new(settings: &ConnectionSettings) -> Self {
+        Self {
+            buffer: Default::default(),
+            first_seq: settings.init_seq_num,
+        }
+    }
+
+    pub fn release_acknowledged_packets(&mut self, acknowledged: SeqNumber) {
+        while acknowledged > self.first_seq {
+            self.buffer.pop_front();
+            self.first_seq += 1;
+        }
+    }
+
+    pub fn get<'a, I: Iterator<Item = SeqNumber> + 'a>(
```

Comment:
A comment describing what this does would be good. Something like iterator of indicies relative to the begin -> data packets